### PR TITLE
Error when integrating SE3 with the same vector given as input and output

### PR DESCRIPTION
**Pull request by duburcqa**
I don't think it would impede performance much since quaternions are fixed-size but you probably know better than me.

### CODE REVIEW

**Review by jcarpent — COMMENTED** [review]
Before digging further, could you provide a reproducible issue example?
This is worth adding as a unit test to enhance the Pinocchio library.

**Comment by duburcqa** [comment]
You can clearly see that without copying the original quaternion in a temporary the chack that is done later after update the quaternion to determine if the sign has been reversed (`const Scalar dot_product = res_quat.dot(quat);`) becomes useless as it always returns exactly `1.0` because `res_quat` and `quat` are precisely pointing to the same memory space :)

**Comment by jcarpent** [comment]
Could you add such an example then in https://github.com/stack-of-tasks/pinocchio/blob/devel/unittest/joint-configurations.cpp? It will allow to not reproduce such bugs in the future.

**Comment by jcarpent** [comment]
Thanks in advance @duburcqa for your help.

**Comment by duburcqa** [comment]
The test [here](https://github.com/duburcqa/pinocchio/blob/d94de14421f6118710bce9e95ab574447fb596af/unittest/liegroups.cpp#L457) is already checking this. Just that random sampling is not a sufficient without many retry. I suggest modifying the test to check many random configuration to be generic. What do you think ?

**Comment by jcarpent** [comment]
This is fine for me to add the randomization of the test to be sure to output a failure case.
Thanks a lot @duburcqa.

**Comment by duburcqa** [comment]
I will have a look at it on sunday I think. I'm quite busy by then.

**Comment by jcarpent** [comment]
@duburcqa Do you have any idea when do you plan to finish this PR?

**Comment by duburcqa** [comment]
I will do it tomorrow. Sorry I have been quite busy lately.

**Comment by jcarpent** [comment]
No problem, thanks a lot for the short notice.

**Comment by duburcqa** [comment]
I updated the unit tests. It fails as expected without the proposed patch. I had to increase the tolerance because it was now failing otherwise due to running evaluating the tests 50 times with different samples now.

File: unittest/liegroups.cpp [hunk 1]
```diff
@@ -91,19 +91,7 @@ void test_lie_group_methods (T & jmodel, typename T::JointDataDerived &)
...
-  {
-    typename T::JointDataDerived jdata = jmodel.createData();
-    jmodel.calc(jdata, qTest);
-    MTest = jdata.M;
-  }
-  {
-    typename T::JointDataDerived jdata = jmodel.createData();
-    jmodel.calc(jdata, qResult);
-    MResult = jdata.M;
-  }
```
Comment (jcarpent) [inline unittest/liegroups.cpp:104]: It would be nice also to keep it.

**Review by jcarpent — APPROVED** [review]
@duburcqa Thanks for finishing this PR.
The PR looks to me, even if I'm surprised by the fact you need to largely increase the tolerance. This should be investigated further in the future.

File: unittest/liegroups.cpp [hunk 2]
```diff
@@ -138,12 +126,12 @@ void test_lie_group_methods (T & jmodel, typename T::JointDataDerived &)
 
   // Check difference
```
Comment (jcarpent) [inline unittest/liegroups.cpp:127]:
```suggestion
  // Check difference
  // TODO(jcarpent): check the increase of tolerance.
```

File: unittest/liegroups.cpp (hunk 1 above)
Comment (jcarpent) [reply unittest/liegroups.cpp:104]: I would say this is complementary ;) always better for coverage.

**Comment by duburcqa** [comment]
I'm also concerned about the updated tolerance thresholds. I could eventually find some time to investigate this issue.

**Comment by jcarpent** [comment]
I've shortly investigated the issue of this PR, and it is, in fact, solved by a more advanced formula present in Pinocchio 3x.

**Review by jcarpent — DISMISSED** [review]

**Review by jcarpent — APPROVED** [review]

**Pull request merged**